File: Algorithms/234. Palindrome Linked List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        if head.is_none() {
            return true;
        }
        
        let mut current = head;
        let mut arr: Vec<i32> = vec![];
        let mut rarr: Vec<i32> = vec![];
        
        while let Some(node) = current {
            arr.push(node.val);
            rarr.push(node.val);
            current = node.next;
        }
        
        rarr.reverse();
        
        arr == rarr
    }
}

File: Algorithms/234. Palindrome Linked List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function isPalindrome(head: ListNode | null): boolean {
  let slow: ListNode = head;
  let fast: ListNode = head;
  const arr: number[] = [];
  while (fast.next && fast.next.next) {
    arr.push(slow.val);
    slow = slow.next;
    fast = fast.next.next;
  }

  if (fast.next) {
    arr.push(slow.val);
  }

  while (slow.next) {
    slow = slow.next;
    if (arr.pop() !== slow.val) {
      return false;
    }
  }
  return true;
}
